src/main/flight/imu.c

```diff
@@ -215,51 +213,79 @@ static float invSqrt(float x)
 STATIC_UNIT_TESTED void imuMahonyAHRSupdate(float dt, float gx, float gy, float gz,
                                 bool useAcc, float ax, float ay, float az,
                                 bool useMag,
-                                float cogYawGain, float courseOverGround, const float dcmKpGain)
+                                float gpsGroundspeedGain, float courseOverGround, const float dcmKpGain)
 {
     static float integralFBx = 0.0f,  integralFBy = 0.0f, integralFBz = 0.0f;    // integral error terms scaled by Ki
 
     // Calculate general spin rate (rad/s)
     const float spin_rate = sqrtf(sq(gx) + sq(gy) + sq(gz));
 
-    // Use raw heading error (from GPS or whatever else)
     float ex = 0, ey = 0, ez = 0;
-    if (cogYawGain != 0.0f) {
-        // Used in a GPS Rescue to boost IMU yaw gain when course over ground and velocity to home differ significantly
 
-        // Compute heading vector in EF from scalar CoG. CoG is clockwise from North
-        // Note that Earth frame X is pointing north and sin/cos argument is anticlockwise
+    // *** Calculate and correct IMU heading from GPS values once min groundspeed exceeded ***
+    if (gpsGroundspeedGain > 0.0f) {
+        // Compute GPS heading vector in earth frame (ef) from scalar GPS courseOverGround
+        // Earth frame X is pointing north and sin/cos argument is anticlockwise
         const fpVector2_t cog_ef = {.x = cos_approx(-courseOverGround), .y = sin_approx(-courseOverGround)};
-#define THRUST_COG 1
-#if THRUST_COG
-        const fpVector2_t heading_ef = {.x = rMat[X][Z], .y = rMat[Y][Z]};  // body Z axis (up) - direction of thrust vector
-#else
-        const fpVector2_t heading_ef = {.x = rMat[0][0], .y = rMat[1][0]};  // body X axis. Projected vector magnitude is reduced as pitch increases
-#endif
-        // cross product = 1 * |heading| * sin(angle) (magnitude of Z vector in 3D)
-        // order operands so that rotation is in direction of zero error
+
+        // Compute IMU / craft heading vector and magnitude in body X axis
+        // Important:  vector magnitude is reduced as pitch increases
+        const fpVector2_t heading_ef = {.x = rMat[0][0], .y = rMat[1][0]};
+        const float headingMagnitude = vector2Mag(&heading_ef);
+
+        // cross (product) = heading * cogHeading * sin error angle, ie, reflects error angle of quad X axis vs GPS groundspeed
+        // cross value sign depends on the rotation change from input to output vectors by right hand rule
+        // operand order is such that rotation is in direction of zero error
+        // abs value of cross is zero when parallel, max of 1.0 at 90 degree error
```

Comment:
> So that, if the nose of the quad was, in the earth plane, directly aligned with the GPS heading, cross would be 1.0, even if the magnitude of the quad vector was attenuated by pitch angulation?

cross == 0 in this case (no error)

gps heading 0, nose pointing east, zero pitch:

 cog_ef  = {1, 0}
 heading_ef  = {0, 1}     (did not test, may be {0, -1} .. still have problems imagining BF coordinate systems. Results will be similar)
cross = -1

gps heading 0, nose pointing east, 60 deg pitch (roll does not have influence here):
 cog_ef  = {1, 0}
 heading_ef  = {0, .5}
cross = -.5


This is actually quite intended for airplane mode - you reduce heading update when flying at steep angle

Reply:
I edited the comment above, I got totally confused, sorry.
Good to confirm that cross is 0 when parallel; that was always my understanding, despite what I wrote, above.  Must have been late at night.